# AI_LV8_CA2_V8

Student: Yuri Ribeiro - 2020347



## Questions


### (i) Identify the differences between a graph and a tree. 

A - A graph is visual representation of information in a more general graphical form than a tree. It consists of a series of nodes/vertices and edges/links displayed as connections linking the nodes. Graphs can have cycles as a closed loops. While a tree is a general data structure similar to a real tree. It also contains nodes/vertices and edges, however, each node has at most one parent node, and none or more children in a specific order. Thus, a tree cannot have any cycles. (Bullinaria, 2019)

###  (ii) Explain in detail how the graph is an abstraction of the problem. 

A- Graph structures are useful for modeling and solving various problems. For example, an underground tunnel network or a railway map can be abstractly represented using nodes and edges. By creating a graph for a railway map, it becomes easier to visualize the paths, distances, and obstacles that can hinder progress, making it simpler to generate new ideas and innovations.(Bullinaria, 2019)

###  (iii)	Identify the advantages of using a visualisation such as the one shown in Fig

A- Using a visualization like fig1 has the advantage of allowing us to identify patterns and structures. It can clarify and create a better understanding of the problem. Also, it is useful to dectect any erros and it helps to design the best algorithm solution.

### (iv)	Demonstrate how Dijkstra’s algorithm would find the shortest path to the solution in Fig.1 through diagrams and written explanation of each stage. 

Basically, the Dijkstra Algorithm starts on the source node that we selected as a start point then it analyses the graph in order to find the shortest path to all nodes ahead. To find the shortest node in an efficient way the algorithm keeps track the distance between every node from the source node and then it updates its values. Once it finds the shortest way the source node and other it marks as a “visited” node and adds it up the value of the way. This process keeps going until all the nodes be visited.

There is a strict requirement for Dijkstra’s Algorithm work: The graphs have to have only positive weights. If negative weight the algorithm will not work properly because negative weights can alter the total weight of visited nodes and decrements after this step has occurred.(Navone, 2020)

In this specific case, the Dijkstra algorithm will work as follows: 

In [1]:
# import image module
from IPython.display import Image

# display the image
Image(url="graph.png")

### Compare the performance of Dijkstra’s algorithm and the A* search algorithm, referring to heuristics, to find the shortest path to the problem by implementing both algorithms 

In [2]:
puzzle_graph = {
    'A' : {'B': 1, 'C' : 5, 'H': 2},
    'B' : {'A': 1, 'C': 2, 'D': 4},
    'C' : {'A': 5, 'B': 2, 'G': 1},
    'D' : {'B': 4, 'E': 4, 'F': 7, 'L': 7},
    'E' : {'D': 4, 'F': 3, 'W': 6},
    'F' : {'D': 7, 'G': 2, 'E': 3 },
    'G' : {'C': 1, 'F': 2, 'L': 3},
    'H' : {'A': 2, 'K': 5, 'J': 9},
    'J' : {'H': 9, 'K': 3, 'N': 6},
    'K' : {'H': 5, 'J': 3, 'L': 5},
    'L' : {'K': 5, 'G': 3, 'D': 7, 'N': 3, 'W': 8, 'M': 4},
    'M' : {'L': 4, 'Q': 10, 'P': 2},
    'N' : {'J': 6, 'L': 3, 'S': 7, 'P': 4},
    'P' : {'N': 4, 'M': 2, 'R': 5},
    'Q' : {'M': 10, 'W': 4, 'S': 8},
    'R' : {'P': 5, 'S': 4, 'T': 3},
    'S' : {'Q': 8, 'R': 4, 'T': 4, 'U':2, 'V': 6},
    'T' : {'R': 3, 'S': 4, 'U': 1},
    'U' : {'S': 2, 'T': 1, 'V': 3},
    'V' : {'W': 5, 'S': 6, 'L': 10, 'U': 3},
    'W' : {'E': 6, 'L': 8, 'Q': 4, 'V': 5}
}

In [26]:
# Dijkstra's Algorithm Performance (source: Week 5 - Artificial Inteligence - CCT - Material)

import heapq
import time

def dijkstra(graph, start):
    # Initialize distances dictionary with infinity for all nodes except the start node
    distances = {node: float('infinity') for node in graph}
    distances[start] = 0  # Distance from start to start is 0

    # Priority queue to store (distance, node) pairs
    priority_queue = [(0, start)]

    # Headp Pop the smallest distance node from the priority queue
    while priority_queue:
        current_distance, current_node = heapq.heappop(priority_queue)

        # Check if the current path is shorter than the known shortest path to the current node
        if current_distance > distances[current_node]:
            continue
            
        # Go to all neighbors nodes of the current node and calculate the total distance from the start node
        for neighbor, weight in graph[current_node].items():
            distance = current_distance + weight

            # If a shorter path is found, update the distance and push it to the priority queue
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(priority_queue, (distance, neighbor))

    return distances

# Set the start and end node
start_node = 'A'
end_node = 'S'

# Record the time when it starts
start_time = time.time()

# variable call the function dijkstra
shortest_distances = dijkstra(puzzle_graph, start_node)

# Record the time when it finisih
end_time = time.time()


# print the result and the time performance
print(f"Shortest distance from node {start_node} to node {end_node}: {shortest_distances[end_node]}")
print(f"Runtime: {round(end_time - start_time, 15)} seconds")

Shortest distance from node A to node S: 17
Runtime: 0.000997543334961 seconds


In [33]:
# A Start Alrogithm Performance( source: https://www.algorithms-and-technologies.com/a_star/python)

import heapq
import math
import time

 # heuristic function returns 0 
def heuristic(node, goal):
   
    return 0 

 # A* algorithm starts from distance 0
    
def a_star(graph, start, goal):
    distances = {node: float('infinity') for node in graph}
    distances[start] = 0
    priority_queue = [(0 + heuristic(start, goal), start)]  # set the cost and node to priority queue

    # Pop the smallest node from the priority queue
    while priority_queue:
        _, current_node = heapq.heappop(priority_queue)

        # go to all neighbors of the current node taken from the priority queue
        for neighbor, weight in graph[current_node].items():
            # calculate the distance from the start node to the neighbor
            distance = distances[current_node] + weight 
            heuristic_cost = heuristic(neighbor, goal)
            total_cost = distance + heuristic_cost

            # If it finds a shorter oath then update the distance and put it on to the priority queue 
            if total_cost < distances[neighbor]:
                distances[neighbor] = total_cost
                heapq.heappush(priority_queue, (total_cost, neighbor))

    return distances

# Set the start and end node
start_node = 'A'
end_node = 'S'

# Record the time when it starts
start_time = time.time()

# variable call the function A_start
shortest_distances_astar = a_star(puzzle_graph, start_node, end_node)

# Record the time when it finisih
end_time = time.time()

# Print the result and the time performance
print(f"Shortest distance from node {start_node} to node {end_node}: {shortest_distances_astar[end_node]}")
print(f"Runtime: {round(end_time - start_time, 10)} seconds")



Shortest distance from node A to node S: 17
Runtime: 0.0009994507 seconds


# Comparison between Dijkstra and A*

In this case, both have the same time complexity: O((V+E). The time complexity of A* depends on the heuristic, however, to solve the puzzle the heuristic was set to 0 becoming A* a Dijkstra's algorithm. 

As we can see the runtime of both performances are:
Dijkstra: 0.000997543334961 seconds
A*:       0.0009994507 seconds

The difference in this case is quite tiny. Making no difference to solve the shortest path problem. However, A* can perform better than Dijkstra in different complex scenarios where heuristics can make a big difference regarding to runtime.

## REFERENCES

Bullinaria, J. (2019). Lecture Notes for Data Structures and Algorithms. Birmingham: School of Computer Science University of Birmingham.


Navone, E. C. (2020, September 28). Dijkstra's Shortest Path Algorithm - A Detailed and Visual Introduction. Retrieved from FreeCode Camp: https://www.freecodecamp.org/news/dijkstras-shortest-path-algorithm-visual-introduction/